packages/react/src/components/ContextMenu/ContextMenu.tsx

```diff
@@ -0,0 +1,234 @@
+
+import {
+  AutoControlledComponent,
+  RenderResultConfig,
+  applyAccessibilityKeyHandlers,
+  getOrGenerateIdFromShorthand,
+  commonPropTypes,
+} from '../../lib'
+import { ShorthandValue, Omit } from '../../types'
+
+import { Accessibility, AccessibilityAttributes } from '../../lib/accessibility/types'
+import { createShorthandFactory } from '../../lib/factories'
+import Popup, { PopupProps } from '../Popup/Popup'
+import { Menu, MenuItemProps, MenuProps, Ref } from '../..'
+import { contextMenuBehavior } from '../../lib/accessibility'
+import { focusMenuItem, focusNearest } from './focusUtils'
+import { ALIGNMENTS, POSITIONS } from '../../lib/positioner'
+
+export interface ContextMenuSlotClassNames {
+  menu: string
+}
+
+export interface ContextMenuProps
+  extends Omit<PopupProps, 'content' | 'renderContent' | 'children' | 'trapFocus'> {
+  /**
+   * Accessibility behavior if overridden by the user.
+   * @default contextMenuBehavior
+   */
+  accessibility?: Accessibility
+
+  menu?: ShorthandValue<MenuProps>
+}
+
+export interface ContextMenuState {
+  open: boolean
+  menuId: string
+  triggerId: string
+  autoFocus: boolean
+}
+
+/**
+ * A ContextMenu displays a menu connected to trigger element.
+ * @accessibility
+ */
+export default class ContextMenu extends AutoControlledComponent<
+  ContextMenuProps,
+  ContextMenuState
+> {
+  static displayName = 'ContextMenu'
+
+  static className = 'ui-contextmenu'
+
+  static create: Function
+
+  static slotClassNames: ContextMenuSlotClassNames = {
+    menu: `${ContextMenu.className}__menu`,
+  }
+
+  static propTypes = {
+    ...commonPropTypes.createCommon({
+      animated: false,
+      as: true,
+      content: false,
+    }),
+    align: PropTypes.oneOf(ALIGNMENTS),
+    defaultOpen: PropTypes.bool,
+    inline: PropTypes.bool,
+    mountDocument: PropTypes.object,
+    mountNode: customPropTypes.domNode,
+    mouseLeaveDelay: PropTypes.number,
+    offset: PropTypes.string,
+    on: PropTypes.oneOfType([
+      PropTypes.oneOf(['hover', 'click', 'focus', 'context']),
+      PropTypes.arrayOf(PropTypes.oneOf(['click', 'focus', 'context'])),
+      PropTypes.arrayOf(PropTypes.oneOf(['hover', 'focus', 'context'])),
+    ]),
+    open: PropTypes.bool,
+    onOpenChange: PropTypes.func,
+    pointing: PropTypes.bool,
+    position: PropTypes.oneOf(POSITIONS),
+    renderContent: PropTypes.func,
+    target: PropTypes.any,
+    trigger: customPropTypes.every([customPropTypes.disallow(['children']), PropTypes.any]),
+    shouldTriggerBeTabbable: PropTypes.bool,
+    unstable_pinned: PropTypes.bool,
+    contentRef: customPropTypes.ref,
+    menu: customPropTypes.itemShorthand,
+  }
+
+  static defaultProps: ContextMenuProps = {
+    accessibility: contextMenuBehavior,
+    align: 'start', // top
+    position: 'below', // after
+  }
+
+  static autoControlledProps = ['open']
+
+  // TODO: this does not persist generated ids across re-renders, see ContextMenu-test.tsx
```

Comment:
@sophieH29 when I use `getOrGenerateIdFromShorthand` here, each re-render will generate new ID. Is that expected or is it a bug?